src/MOI_wrapper.jl

```diff
@@ -255,31 +266,44 @@ Note that this method *does not returns* the actual jacobians.
 For more info refer eqn(7) and eqn(8) of https://arxiv.org/pdf/1703.00443.pdf
 """
 function backward!(model::Optimizer, params::Vector{String}, dl_dz::Vector{Float64})
-    Q, q, G, h, A, b, nz, var_idx, nineq, ineq_con_idx, neq, eq_con_idx = get_problem_data(model.optimizer)
-
     z = model.primal_optimal
-
-    # separate λ, ν
-    λ = filter(con -> !is_equality(MOI.get(model.optimizer, MOI.ConstraintSet(), con)), model.con_idx)
-    ν = filter(con -> is_equality(MOI.get(model.optimizer, MOI.ConstraintSet(), con)), model.con_idx)
-
-    λ = [MOI.get(model.optimizer, MOI.ConstraintDual(), con) for con in λ]
-    ν = [MOI.get(model.optimizer, MOI.ConstraintDual(), con) for con in ν]
-
-    grads = []
-    LHS = create_LHS_matrix(z, λ, Q, G, h, A)
+    if !isempty(model.gradient_cache)
+        (Q, q, G, h, A, b, nz, var_idx, nineq, ineq_con_idx, neq, eq_con_idx) = model.gradient_cache.problem_data
+        λ = model.gradient_cache.inequality_duals
+        ν = model.gradient_cache.equality_duals
+        LHS = model.gradient_cache.lhs
+
+    else # full computation
+        Q, q, G, h, A, b, nz, var_idx, nineq, ineq_con_idx, neq, eq_con_idx = get_problem_data(model.optimizer)
+
+        # separate λ, ν
+        λ = filter(con -> !is_equality(MOI.get(model.optimizer, MOI.ConstraintSet(), con)), model.con_idx)
+        ν = filter(con -> is_equality(MOI.get(model.optimizer, MOI.ConstraintSet(), con)), model.con_idx)
+
+        λ = [MOI.get(model.optimizer, MOI.ConstraintDual(), con) for con in λ]
+        ν = [MOI.get(model.optimizer, MOI.ConstraintDual(), con) for con in ν]
+
+        LHS = create_LHS_matrix(z, λ, Q, G, h, A)
+        model.gradient_cache = (
+            problem_data = (Q, q, G, h, A, b, nz, var_idx, nineq, ineq_con_idx, neq, eq_con_idx),
+            inequality_duals = λ,
+            equality_duals = ν,
+            lhs = LHS,
+        )
+    end
+    
     RHS = [dl_dz; zeros(neq+nineq)]
-
     partial_grads = -(LHS \ RHS)
-
     dz = partial_grads[1:nz]
+
     if nineq > 0
         dλ = partial_grads[nz+1:nz+nineq]
     end
     if neq > 0
         dν = partial_grads[nz+nineq+1:nz+nineq+neq]
     end
 
+    grads = []
```

Comment:
Can't we have at least `Union{Vector{Float64}, Matrix{Float64}}[]` ? It's not only for perf, it also helps readability.

Reply:
This is just a line that was moved from another place in the function, I plan to return gradients in a cleaner way in another one